internal/server/server.go

```diff
@@ -1,24 +1,168 @@
-	router *http.ServeMux
+	router                 *http.ServeMux
+	cookieName             string
+	groups                 config.Groups
+	grafanaProxyUrl        *url.URL
+	grafanaClient          *grafana.Client
+	grafanaResponseHeaders GrafanaResponseHeaders
+	skipTLSVerify          bool
 }
 
-func New() http.Handler {
+type ServerFuncOpt func(*Server) error
+
+func New(opts ...ServerFuncOpt) (http.Handler, error) {
 	s := &Server{router: http.NewServeMux()}
 
+	// load options
+	for _, opt := range opts {
+		err := opt(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	s.router.HandleFunc("/", s.handleRoot())
 
-	return s.router
+	return s.router, nil
 }
 
 func (s *Server) handleRoot() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintln(w, "hello world")
+		// Extract token from cookie
+		cookie, err := r.Cookie(s.cookieName)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading cookie")
+			return
+		}
+
+		// Get claims from token
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		if claims.Subject == "" {
+			logAndError(w, http.StatusUnauthorized, err, "sub claim is empty")
+			return
+		}
+
+		if len(claims.Groups) == 0 {
+			logAndError(w, http.StatusUnauthorized, err, "groups claim is empty")
+			return
+		}
+
+		// Make Subject claim the value used for login
+		login := claims.Subject
+
+		log.Infof("user %s is attempting to log in", login)
+		log.Debugf("claim groups for user %s: %v", login, claims.Groups)
+
+		// validUserGroups represents the intersection of user groups from claim with the group
+		// mapping in configuration
+		validUserGroups := config.ValidUserGroups(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", login, validUserGroups)
+		if len(validUserGroups) == 0 {
+			logAndError(w, http.StatusUnauthorized, err, "no user groups matching configured mapping")
+			return
+		}
+
+		// lookup the user globally first as if it is not present it would need to
+		// be created
+		orgUser, err := s.grafanaClient.LookupUser(claims.Subject)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error looking for user")
+			return
+		}
+
+		// Assign the email from claim so it's always updated
+		orgUser.Email = claims.Email
+
+		// if the Login field in user is empty, it means that the user wasn't found
+		if orgUser.Login == "" {
+			log.Infof("no user with login %s found, creating new one", login)
+			orgUser.Login = login
+
+			uid, err := s.grafanaClient.CreateUser(orgUser)
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error creating new user")
+				return
+			}
+
+			orgUser.ID = uid
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
+
+		for _, groupConfig := range validUserGroups {
+			if groupConfig.GrafanaAdmin && !orgUser.IsAdmin {
```

Comment:
This seem like we will not be able to revoke GrafanaAdmin in configuration.

Reply:
fixed